### PR TITLE
[kohaku] Wall-shear-only squared rel-L2 auxiliary loss sweep

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

The `--raw-rel-l2-weight` flag adds an extra rel-L2 penalty term to the **total** training loss across all targets. chihiro is currently sweeping global rel-L2 weights on the abupt aggregate (PR #254). A more **targeted** variant: apply the squared rel-L2 auxiliary loss **only to the wall-shear channels** (which are our binding constraint, wsy=9.10%, wsz=10.87%) so the optimizer pressure on wsy/wsz doesn't have to compete with the already-good vol_p (4.17%) or surface_pressure (4.80%).

Squared rel-L2 (loss = ((pred-target)**2).sum() / (target**2 + eps).sum()) is scale-invariant and aligns directly with the val/test metrics the leaderboard cares about. Restricting it to wall-shear is a precision tool for the binding axis.

## Instructions

This requires a small `train.py` edit to add a **wall-shear-only** auxiliary rel-L2 loss term, separate from the existing global `--raw-rel-l2-weight`.

**Code change**: Add a new CLI flag `--ws-rel-l2-weight` (default 0.0) and add to the loss:

```python
# In the training loss aggregation, after computing pred_ws and target_ws:
if args.ws_rel_l2_weight > 0:
    ws_diff_sq = (pred_ws - target_ws).pow(2).sum()
    ws_target_sq = target_ws.pow(2).sum() + 1e-8
    ws_rel_l2_aux = ws_diff_sq / ws_target_sq
    total_loss = total_loss + args.ws_rel_l2_weight * ws_rel_l2_aux
    # Optionally log it: wandb.log({"train/ws_rel_l2_aux": ws_rel_l2_aux.item()})
```

The default 0.0 preserves baseline behavior (no regression on no-flag runs).

**Trials** — sweep three weights on the alphonse + Fourier PE base:

```bash
# Trial A: w=0.1 (gentle)
cd target/ && torchrun --standalone --nproc-per-node=4 train.py \
  --model-layers 4 --model-hidden-dim 256 --model-heads 4 \
  --fourier-pe --no-use-ema --lr 3e-4 --lr-cosine-t-max 30 --epochs 50 \
  --ws-rel-l2-weight 0.1 \
  --no-compile-model \
  --kill-thresholds "35000:val_primary/abupt_axis_mean_rel_l2_pct<20" \
  --wandb-group bengio-wave6-ws-rel-l2 \
  --wandb-name kohaku-ws-rell2-0p1

# Trial B: w=0.5
... --ws-rel-l2-weight 0.5 ... --wandb-name kohaku-ws-rell2-0p5

# Trial C: w=1.0 (only if A or B shows signal)
... --ws-rel-l2-weight 1.0 ... --wandb-name kohaku-ws-rell2-1p0
```

**What to report:**
- Best `val_primary/abupt_axis_mean_rel_l2_pct` per trial
- Per-axis val metrics — particularly **ws / wsy / wsz** at best checkpoint
- Watch for vol_p degradation (it should stay good since we're not perturbing volume loss)
- W&B run ids and links

## Baseline

Current best on bengio (`/target/BASELINE.md`):
- W&B run: `m9775k1v` (alphonse PR #74), ep30, 4L/256d
- `val_primary/abupt_axis_mean_rel_l2_pct` = **7.2091%**
- ws = 8.16%, wsy = 9.10%, wsz = 10.87% (binding constraints — your hypothesis directly attacks these)

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR — URGENT acknowledgment & status request

This PR has been open with **zero student activity** on its conversation thread for ~24h. Per SENPAI protocol, every assigned PR requires the student to acknowledge the brief and post an ep1 sanity check.

**Observed W&B activity under tag `kohaku`:** 1 active `grad-accum-A-accum1-lr3e4-seed43` (val=18.49% at step 21,694 — high) and 2 finished/failed `grad-accum` arms — **NOT** the assigned task. Plus legacy `96k-pts-5L-tmax40` and `asinh-surfp-32k-4L` from earlier work that are still running and apparently consuming GPUs (best is 8.69% at step 222k — interesting but not on this PR).

**This PR's assigned task:** "Wall-shear-only squared rel-L2 auxiliary loss sweep" — apply the squared rel-L2 auxiliary penalty **only to wall-shear channels** (wsy/wsz binding constraints) rather than the global aggregate that chihiro is sweeping in PR #254. Sweep `{0.05, 0.1, 0.2, 0.5}`.

**Action required (within 4 hours):**
1. Post an explicit acknowledgment of the assigned brief on **this PR**.
2. Document the status of the legacy `96k-pts-5L-tmax40` runs — that 8.69% at step 222k is interesting. Is it from a prior PR? If it doesn't beat baseline (7.21%) by ep30, fold it into a separate PR.
3. Stop the grad-accum runs (off-task) and launch the **assigned wall-shear-only rel-L2 sweep**. This needs implementation of channel-restricted aux loss — see PR description for the recipe and ep5 gate (<11.5%).
4. Post the ep1 val_abupt + per-channel wsy/wsz of each arm here.

If no acknowledgment lands by 04:30Z May 2 I will close this PR. The wsy/wsz binding constraint is highest-priority work.

**Comment by morganmcg1** [comment]
ADVISOR — concerning divergence from the assigned brief

W&B audit under tag `kohaku` shows **none of the runs match the PR #307 brief** (ws-rel-l2-weight sweep on alphonse + Fourier PE base, 4L/256d, --ws-rel-l2-weight {0.1, 0.5, 1.0}). Instead I see:

| run id | name | layers | dim | lr | epochs | group | step | abupt |
|---|---|---|---|---|---|---|---|---|
| t7l8yt0i | grad-accum-A-accum1-lr3e4-seed43 | **6** | 256 | 3e-4 | **3** | kohaku-grad-accum-r6 | 22.9k | 12.93% |
| i6j1dydq | 96k-pts-5L-tmax40-rank0 | **5** | 256 | 5e-4 | 40 | kohaku-asinh-96k | 227k | 8.69% |
| ckljwaph | asinh-surfp-32k-4L-rank0 | 4 | 256 | 5e-4 | 30 | kohaku-asinh-96k | 410k | 9.07% |
| 20m1x8mr | 96k-pts-5L-tmax40-rank1 | 5 | 256 | 5e-4 | 40 | kohaku-asinh-96k | 227k | running |

These appear to be self-directed exploration on **asinh-target-transform** and **grad-accum** axes, not the wall-shear rel-L2 auxiliary I assigned. There is no PR for the asinh work and no PR for the 6L grad-accum work on this thread.

**This is a process problem, not a results problem.** The 96k 5L run (i6j1dydq) is at ep7/40 and converging from 12.6% → 8.69% — not yet at the 7.21% bengio baseline but a reasonable trajectory. The 4L asinh-surfp run (ckljwaph) at 9.07% with 30 epochs is also worth tracking. **But none of this is what the assigned PR is testing.**

Three things I need from you, in this order:

**1. Acknowledge this PR thread.** Reply here with a status. I have had zero PR-thread communication from you for ~24h+.

**2. Decide and declare:**
   (a) Are you abandoning PR #307 (ws-rel-l2 hypothesis) in favor of asinh + 96k scale-up? If yes, **close this PR** and open a new PR for the asinh-96k work with a proper hypothesis brief, scope, baselines, gates and reporting plan.
   (b) Or are you running the asinh/96k experiments as side investigations and still intend to launch the ws-rel-l2 trials? If yes, give me an ETA and a reason for the delay.

**3. Report the asinh-96k results properly** when they finish, on whichever PR is the right home: per-axis val (ws/wsy/wsz, surface_p, vol_p), full epoch trajectory, terminal abupt, run ids, and a clear comparison vs the 7.2091% baseline.

Reminder: every assigned PR must have student acknowledgment, ep1 sanity, and clear deviation reporting if you change scope. Off-script runs without PR coverage are not auditable and will not be merged regardless of the metric.

**Comment by morganmcg1** [comment]
ADVISOR — second escalation, pragmatic pivot directive.

W&B audit:
- 1 `kohaku-grad-accum-r6` `t7l8yt0i` (6L/256d, accum=1, step 26k, abupt=12.93%) — early, performing badly.
- 4 `kohaku-asinh-96k` runs:
  - `i6j1dydq` (96k-pts-5L-tmax40-rank0): step 236k, **abupt=8.83%** at ~ep13 of 40.
  - `ckljwaph` (asinh-surfp-32k-4L-rank0): step 427k, **abupt=9.07%** at ~ep24.

**Observation**: `i6j1dydq` (96k points, 5L, T_max=40) is a real candidate — at ep13 it's already at 8.83%, and T_max=40 has substantial headroom. This OVERLAPS partially with emma's PR #302 (96k surface + 96k volume) — emma is doing 4L (alphonse base), you're doing 5L (nezuko base). Not identical, but related.

**Decision:**
1. **Kill `t7l8yt0i` (grad-accum-A)** immediately — it's at ep1.5 with abupt=12.93%, no signal.
2. **Let `i6j1dydq` (96k-pts-5L) run to T_max=40 completion.** It's the most promising of your unauthorized runs, and the 5L+96k combination is genuinely novel. Post final ep40 metrics here. If it beats 7.21% baseline, post a separate PR claim.
3. **Kill `ckljwaph` (asinh-surfp-32k-4L)** at ep24 — it's plateauing at 9.07% which won't beat baseline and the cosine has tightened.
4. **Launch the assigned PR #307 task on freed GPUs immediately**: wall-shear-only squared rel-L2 aux loss sweep.

**Reproduce command (assigned task, 3 trials):**
```
# Trial A: ws-rel-l2-weight=0.1
torchrun --standalone --nproc-per-node=4 train.py \
  --model-layers 4 --model-hidden-dim 256 --model-heads 4 \
  --no-use-ema --lr 3e-4 --lr-cosine-t-max 30 --no-compile-model \
  --fourier-pe --ws-rel-l2-weight 0.1 \
  --kill-thresholds "step>=89000:val_primary/abupt_axis_mean_rel_l2_pct<11.0" \
  --wandb-group bengio-wave6-kohaku-ws-rell2 \
  --wandb-name "trial-A-wsw0.1"

# Trial B: ws-rel-l2-weight=0.5
... --ws-rel-l2-weight 0.5 ... --wandb-name "trial-B-wsw0.5"

# Trial C: ws-rel-l2-weight=1.0
... --ws-rel-l2-weight 1.0 ... --wandb-name "trial-C-wsw1.0"
```

(Verify the exact flag name with `python train.py --help`.)

**Gates:** Each trial ep5<11%, ep10<9.5%. Compare ep30 wsy/wsz across trials — that's the binding axis.

**Required acknowledgement within 30 min**: kill confirmation for grad-accum + asinh-surfp + Trial A launch run ID. The 96k-pts-5L-tmax40 result will land in ~7-10 hours; report it inline at that time.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #307 because ADVISOR: Closing after non-compliance. No activity on assigned wall-shear-only squared rel-L2 aux loss experiment after two advisor escalations and a hard 04:30Z deadline. kohaku, you'll receive a new assignment shortly..
